src/core/lb/particle_coupling.hpp

```diff
@@ -83,79 +58,56 @@ void add_md_force(LB::Solver &lb, Utils::Vector3d const &pos,
 // internal function exposed for unit testing
 std::vector<Utils::Vector3d> positions_in_halo(Utils::Vector3d const &pos,
                                                BoxGeometry const &box,
+                                               LocalBox const &local_geo,
                                                double agrid);
 
-// internal function exposed for unit testing
-void add_swimmer_force(Particle const &p, double time_step);
-
-void mpi_bcast_lb_particle_coupling();
-
 /** @brief Calculate drag force on a single particle.
  *
  *  See section II.C. @cite ahlrichs99a
  *
  *  @param[in] lb          The coupled fluid
+ *  @param[in] lb_gamma    The friction coefficient
  *  @param[in] p           The coupled particle
  *  @param[in] shifted_pos The particle position in LB units with optional shift
  *  @param[in] vel_offset  Velocity offset in MD units to be added to
  *                         interpolated LB velocity before calculating the force
  *
  *  @return The viscous coupling force
  */
-Utils::Vector3d lb_drag_force(LB::Solver const &lb, Particle const &p,
+Utils::Vector3d lb_drag_force(LB::Solver const &lb, double lb_gamma,
+                              Particle const &p,
                               Utils::Vector3d const &shifted_pos,
                               Utils::Vector3d const &vel_offset);
 
-namespace LB {
-struct ParticleCouplingConfig {
-  OptionalCounter rng_counter_coupling = {};
-  /** @brief Friction coefficient for the particle coupling. */
-  double gamma = 0.0;
-  bool couple_to_md = false;
-
-private:
-  friend class boost::serialization::access;
-
-  template <class Archive> void serialize(Archive &ar, const unsigned int) {
-    ar &rng_counter_coupling;
-    ar &gamma;
-    ar &couple_to_md;
-  }
-};
-} // namespace LB
-
-// internal global exposed for unit testing
-extern LB::ParticleCouplingConfig lb_particle_coupling;
-
 namespace LB {
 
-/** @brief Calculate particle-lattice interactions. */
-void couple_particles(ParticleRange const &real_particles,
-                      ParticleRange const &ghost_particles, double time_step);
-
 class ParticleCoupling {
+  LBThermostat const &m_thermostat;
   LB::Solver &m_lb;
-  bool m_thermalized;
+  BoxGeometry const &m_box_geo;
+  LocalBox const &m_local_box;
   double m_time_step;
   double m_noise_pref_wo_gamma;
+  bool m_thermalized;
 
 public:
-  ParticleCoupling(LB::Solver &lb, double time_step, double kT)
-      : m_lb{lb}, m_thermalized{kT != 0.}, m_time_step{time_step} {
-    assert(kT >= 0.);
+  ParticleCoupling(LBThermostat const &thermostat, LB::Solver &lb,
+                   BoxGeometry const &box_geo, LocalBox const &local_box,
+                   double time_step, double kT = -1.)
```

Comment:
Is this really a good design choice? I'm not a fan of "changing the coupling behaviour" based on an invalid default argument... However, if it's the best choice for the moment, leave it as it is.

Reply:
I agree it's not great, but I wanted to avoid having two constructors. The core uses the default constructor with a negative temperature and gets the actual temperature from the LB fluid. The unit test passes a custom temperature to verify particle coupling is properly thermalized. I'll keep that in mind when rewriting the unit tests as part of #4734 and look at the feasibility of instantiating a LB fluid with a different temperature in every test to remove this default argument.

Reply:
A much cleaner solution was now pushed to this PR.